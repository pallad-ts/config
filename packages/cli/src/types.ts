import { Either } from "@sweet-monads/either";

import { Provider } from "@pallad/config";

export interface LoadedConfig {
    config: unknown;
    providerMap: Map<Provider<unknown>, Provider.Value<unknown>>;
}
