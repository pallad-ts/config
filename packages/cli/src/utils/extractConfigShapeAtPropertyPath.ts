import { Either, left, right } from "@sweet-monads/either";

import { Provider } from "@pallad/config";

import { ERRORS } from "../errors";

export async function extractConfigShapeAtPropertyPath(
    config: unknown,
    propertyPath: string
): Promise<Either<Error | Provider.Fail, unknown>> {
    let currentConfig = config;
    const currentPropertyPath = [];
    for (const part of propertyPath.split(".")) {
        currentPropertyPath.push(part);
        if (!isValidConfigValue(currentConfig)) {
            return left(ERRORS.INVALID_CONFIG_VALUE_AT_PATH.create(currentConfig, currentPropertyPath));
        }

        let newCurrentConfig = (currentConfig as any)[part];

        if (Provider.isType(newCurrentConfig)) {
            const resolvedProviderValue = await newCurrentConfig.getValue();

            if (resolvedProviderValue.isLeft()) {
                return left(resolvedProviderValue.value);
            }

            newCurrentConfig = resolvedProviderValue.value;
        }
        currentConfig = newCurrentConfig;
    }

    return right(currentConfig);
}

function isValidConfigValue(value: unknown): value is object | unknown[] {
    // eslint-disable-next-line no-null/no-null
    return value !== null && value !== undefined && typeof value !== "function";
}
