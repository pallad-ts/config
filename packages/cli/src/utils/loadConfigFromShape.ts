import * as is from "predicates";

import { extractProvidersFromConfig, Provider, replaceProvidersInConfig } from "@pallad/config";

import { LoadedConfig } from "../types";

export async function loadConfigFromShape(configShape: unknown): Promise<LoadedConfig> {
    const map = new Map<Provider<any>, Provider.Value<any>>();
    if (is.primitive(configShape)) {
        return { config: configShape, providerMap: map };
    }
    const providers = Array.from(extractProvidersFromConfig(configShape));

    await Promise.all(
        providers.map(async provider => {
            const value = await provider.getValue();

            map.set(provider, value);
        })
    );

    return {
        config: replaceProvidersInConfig(configShape, map as any),
        providerMap: map,
    };
}
