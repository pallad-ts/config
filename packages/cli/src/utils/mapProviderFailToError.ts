import { ERRORS, Provider, ValueNotAvailable } from "@pallad/config";

export function mapProviderFailToError(value: Provider.Fail | Provider.Fail[]): Error | Error[] {
    return Array.isArray(value) ? value.map(mapSingleFail) : mapSingleFail(value);
}

function mapSingleFail(value: Provider.Fail): Error {
    if (ValueNotAvailable.isType(value)) {
        return ERRORS.PROVIDER_VALUE_NOT_AVAILABLE.create(value.description);
    }
    return value;
}
