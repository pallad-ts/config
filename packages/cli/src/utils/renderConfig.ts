import { isEither } from "@sweet-monads/either";
import chalk from "chalk";
import { Config, format as prettyFormat, Plugin, Printer, Refs } from "pretty-format";

import { ERRORS, Provider, ValueNotAvailable } from "@pallad/config";
import { Secret } from "@pallad/secret";

export function renderConfig(config: unknown, options: displayConfig.Options) {
    function renderSecret(secret: Secret<any>, printer: (value: any) => string) {
        if (options.revealSecrets) {
            return printer(secret.getValue());
        }
        return secret.getDescription();
    }

    const secretPlugin: Plugin = {
        test(value: any) {
            return Secret.is(value);
        },
        serialize(val: Secret<any>, config: Config, indentation: string, depth: number, refs: Refs, printer: Printer) {
            return (
                renderSecret(val, value => printer(value, config, indentation, depth, refs, true)) +
                " " +
                chalk.yellow("(secret)")
            );
        },
    };

    const providerPlugin: Plugin = {
        test(value: any) {
            return value && isEither(value);
        },
        serialize(
            val: Provider.Value<any>,
            config: Config,
            indentation: string,
            depth: number,
            refs: Refs,
            printer: Printer
        ) {
            if (val.isRight()) {
                return printer(val.value, config, indentation, depth, refs, false);
            }
            const fail = val.value;
            const errors = (Array.isArray(fail) ? fail : [fail])
                .map(x => {
                    if (ValueNotAvailable.isType(x)) {
                        return ERRORS.PROVIDER_VALUE_NOT_AVAILABLE.create(x.description);
                    }
                    return x;
                })
                .map(x => x.message);
            return chalk.red(errors.join(", "));
        },
    };

    return prettyFormat(config, {
        plugins: [secretPlugin, providerPlugin],
    });
}

export namespace displayConfig {
    export interface Options {
        revealSecrets: boolean;
    }
}
