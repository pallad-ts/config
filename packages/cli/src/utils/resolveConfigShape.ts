import { Either, fromPromise } from "@sweet-monads/either";
import { realpath } from "node:fs/promises";
import { pathToFileURL } from "node:url";

import { validateConfigSource } from "./validateConfigSource";

export async function resolveConfigShape(options: resolveConfigShape.Options): Promise<Either<Error, unknown>> {
    return (await fromPromise<Error, unknown>(loadModule(options.filePath))).asyncChain(module => {
        return validateConfigSource(extractConfigFromModule(module, options.moduleConfigProperty));
    });
}

export namespace resolveConfigShape {
    export interface Options {
        filePath: string;
        moduleConfigProperty?: string;
    }
}

// this is highly inspired from https://github.com/cosmiconfig/cosmiconfig/blob/main/src/loaders.ts
async function loadModule(modulePath: string) {
    const realModulePath = await realpath(modulePath);
    try {
        const fileUrl = pathToFileURL(realModulePath);
        return await import(fileUrl.href);
    } catch (error) {
        try {
            return require(realModulePath);
        } catch (requireError: unknown) {
            if (
                (requireError instanceof Error && (requireError as any).code === "ERR_REQUIRE_ESM") ||
                (requireError instanceof SyntaxError &&
                    requireError.toString().includes("Cannot use import statement outside a module"))
            ) {
                throw error;
            }

            throw requireError;
        }
    }
}

function extractConfigFromModule(module: any, property?: string) {
    if (property) {
        return module[property];
    }
    if (module instanceof Function) {
        return module;
    }
    if ("default" in module) {
        return module.default;
    }
}
