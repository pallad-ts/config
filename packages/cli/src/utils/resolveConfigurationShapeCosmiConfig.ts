import { Either, left, right } from "@sweet-monads/either";
import { fromNullable, Maybe } from "@sweet-monads/maybe";
import { PublicExplorer as CosmiConfigPublicExplorer } from "cosmiconfig";
import * as is from "predicates";

export async function resolveConfigurationShapeCosmiConfig(
    explorer: CosmiConfigPublicExplorer,
    { filePath, directoryStartPath }: resolveConfigurationShapeCosmiConfig.Options
): Promise<Maybe<Either<Error, unknown>>> {
    const result = await (filePath ? explorer.load(filePath) : explorer.search(directoryStartPath));

    return fromNullable(result?.config).asyncMap(validateConfig);
}

async function validateConfig(config: unknown): Promise<Either<Error, unknown>> {
    if (config instanceof Function) {
        return validateConfigShape(await config());
    }

    return validateConfigShape(config);
}

function validateConfigShape(shape: unknown): Either<Error, unknown> {
    if (is.function(shape)) {
        left(new Error("Configuration shape must be an object, primitive or array. Got " + typeof shape));
    }

    return right(shape);
}

export namespace resolveConfigurationShapeCosmiConfig {
    export interface Options {
        filePath?: string;
        directoryStartPath?: string;
    }
}
