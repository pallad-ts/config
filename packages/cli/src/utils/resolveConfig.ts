import { Either } from "@sweet-monads/either";

import { LoadedConfig } from "../types";
import { loadConfigFromShape } from "./loadConfigFromShape";
import { resolveConfigShape } from "./resolveConfigShape";

export async function resolveConfig(options: resolveConfigShape.Options): Promise<Either<Error, LoadedConfig>> {
    return (await resolveConfigShape(options)).asyncMap(configShape => {
        return loadConfigFromShape(configShape);
    });
}
