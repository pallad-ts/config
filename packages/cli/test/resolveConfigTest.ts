import { LoadedConfig } from "@src/types";
import { resolveConfigShape } from "@src/utils/resolveConfigShape";
import { Either } from "@sweet-monads/either";
import { Maybe } from "@sweet-monads/maybe";
import { AssertionError } from "node:assert";
import * as path from "node:path";

function assertConfigIsFound(config: Maybe<Either<Error, LoadedConfig>>): Either<Error, LoadedConfig> {
    if (config.isJust()) {
        return config.value;
    }

    throw new AssertionError({ message: "Config not found" });
}

function assertConfigHasNoErrors(config: Either<Error, LoadedConfig>): LoadedConfig {
    if (config.isRight()) {
        return config.value;
    }

    throw new AssertionError({ message: "Config has errors" });
}

describe("resolveConfig", () => {
    const FIXTURES_DIR = path.join(__dirname, "./fixtures");

    it.each([["object"], ["function"], ["async-function"], ["primitive"]])("resolving as %s", async type => {
        const result = await resolveConfigShape({
            directoryStartPath: path.join(FIXTURES_DIR, `config-as-${type}`),
        });

        const config = assertConfigHasNoErrors(assertConfigIsFound(result));

        expect(config.config).toMatchSnapshot();
    });

    it("loading config from file path", async () => {
        const result = await resolveConfigShape({
            filePath: path.join(FIXTURES_DIR, `custom-config-path/custom-config.ts`),
        });

        const config = assertConfigHasNoErrors(assertConfigIsFound(result));

        expect(config.config).toMatchSnapshot();
    });

    describe("fails", () => {
        it("if config is not found", async () => {
            const result = await resolveConfigShape({
                directoryStartPath: path.join(FIXTURES_DIR, `config-not-found`),
            });

            expect(result.isNone()).toBe(true);
        });

        it("if config returns a function", async () => {
            const result = await resolveConfigShape({
                directoryStartPath: path.join(FIXTURES_DIR, `invalid-config`),
            });

            expect(result.isJust()).toBe(true);
            expect(result.value!.value).toMatchSnapshot();
        });
    });
});
