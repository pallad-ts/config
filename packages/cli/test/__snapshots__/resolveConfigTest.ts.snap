// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`resolveConfig fails if config returns a function 1`] = `
{
  "config": [Function],
  "providerMap": Map {},
}
`;

exports[`resolveConfig loading config from file path 1`] = `
{
  "database": {
    "hostname": "hostname",
    "password": **Secret value: "database password" **,
    "port": 5432,
    "schema": "test",
    "username": **Secret value: "username" **,
  },
  "jwt": {
    "secret": **Secret value: [
      "s1",
      "s2",
      "s3",
    ] **,
  },
  "scheduler": {
    "options": {
      "dns": "http://localhost:8080",
    },
    "type": "dynamic",
  },
  "usernames": {
    "fake": [
      "tommy",
      "john",
      "jerry",
    ],
  },
}
`;

exports[`resolveConfig resolving as async-function 1`] = `
{
  "database": {
    "hostname": "hostname",
    "password": **Secret value: "database password" **,
    "port": 5432,
    "schema": "test",
    "username": **Secret value: "username" **,
  },
  "jwt": {
    "secret": **Secret value: [
      "s1",
      "s2",
      "s3",
    ] **,
  },
  "scheduler": {
    "options": {
      "dns": "http://localhost:8080",
    },
    "type": "dynamic",
  },
  "usernames": {
    "fake": [
      "tommy",
      "john",
      "jerry",
    ],
  },
}
`;

exports[`resolveConfig resolving as function 1`] = `
{
  "database": {
    "hostname": "hostname",
    "password": **Secret value: "database password" **,
    "port": 5432,
    "schema": "test",
    "username": **Secret value: "username" **,
  },
  "jwt": {
    "secret": **Secret value: [
      "s1",
      "s2",
      "s3",
    ] **,
  },
  "scheduler": {
    "options": {
      "dns": "http://localhost:8080",
    },
    "type": "dynamic",
  },
  "usernames": {
    "fake": [
      "tommy",
      "john",
      "jerry",
    ],
  },
}
`;

exports[`resolveConfig resolving as object 1`] = `
{
  "database": {
    "hostname": "hostname",
    "password": **Secret value: "database password" **,
    "port": 5432,
    "schema": "test",
    "username": **Secret value: "username" **,
  },
  "jwt": {
    "secret": **Secret value: [
      "s1",
      "s2",
      "s3",
    ] **,
  },
  "scheduler": {
    "options": {
      "dns": "http://localhost:8080",
    },
    "type": "dynamic",
  },
  "usernames": {
    "fake": [
      "tommy",
      "john",
      "jerry",
    ],
  },
}
`;

exports[`resolveConfig resolving as primitive 1`] = `"some primitive value"`;
